src/passes/OptimizeInstructions.cpp

```diff
@@ -1447,6 +1473,40 @@ struct OptimizeInstructions
     }
   }
 
+  void visitRefEq(RefEq* curr) {
+    // Equality does not depend on the type, so casts may be removable.
+    //
+    // This is safe to do first because nothing else here cares about the type,
+    // and we consume the two input references, so removing a cast could not
+    // help our parents (see "notes on removing casts").
+    skipCast(curr->left, Type::eqref);
+    skipCast(curr->right, Type::eqref);
```

Comment:
Do we need the condition of subtypes of `eqref`? If either of them is not a subtype of `eqref`, does this even validate?

Reply:
The tricky situation is something like this:
```wat
(ref.eq
  (ref.cast $A
    (ref.func $func)
  ..
)
```
That will trap at runtime. But at compile time we can't remove the cast, as it would fail to validate - but it does validate before the removal.